### PR TITLE
Order API's Listing, Searching, Sorting & Query Helper Methods

### PR DESCRIPTION
**Pull request by shewa12**

### CODE REVIEW

File: ecommerce/OrderController.php [hunk 1]
```diff
@@ -48,38 +64,100 @@ class OrderController {
...
+	 * If the order ID is not provided, it responds with a "Bad Request" status.
+	 * If the order is not found, it responds with a "Not Found" status.
+	 * Otherwise, it responds with the order data and a success message.
+	 *
+	 * @since 3.0.0
+	 *
+	 * @return void
+	 */
+	public function get_order_by_id() {
+
```
Comment (harunollyo) [inline ecommerce/OrderController.php:129]: Nonce verification required

```php
if ( ! tutor_utils()->is_nonce_verified() ) {
     $this->json_response( tutor_utils()->error_message( 'nonce' ), null, HttpHelper::STATUS_BAD_REQUEST );
}
```

**Review by harunollyo — CHANGES_REQUESTED** [review]

File: assets/react/admin-dashboard/segments/filter.js [hunk 2]
```diff
@@ -43,6 +43,16 @@ document.addEventListener('DOMContentLoaded', function () {
...
+	const filterPaymentStatus = document.getElementById('tutor-backend-filter-payment-status');
+	if (filterPaymentStatus) {
+		filterPaymentStatus.addEventListener(
+			'change',
+			(e) => {
+				window.location = urlPrams('payment-status', e.target.value);
+			},
+			{ once: true },
+		);
+	}
```
Comment (ahamed) [inline assets/react/admin-dashboard/segments/filter.js:55]: You can remove the if condition by using optional chaining.
```suggestion
		filterPaymentStatus?.addEventListener(
			'change',
			(e) => {
				window.location = urlPrams('payment-status', e.target.value);
			},
			{ once: true },
		);
```

BTW, why does this event is attached for once?

File: ecommerce/OrderController.php [hunk 3]
```diff
@@ -2,37 +2,53 @@
...
+ * @since 3.0.0
  */
 class OrderController {
+
+	/**
+	 * Order model
+	 *
+	 * @since 3.0.0
+	 *
+	 * @var Object
```
Comment (ahamed) [inline ecommerce/OrderController.php:36]:
```suggestion
	 * @var OrderModel
```

File: ecommerce/OrderController.php [hunk 4]
```diff
@@ -103,77 +181,49 @@ public function prepare_bulk_actions(): array {
...
-			),
-			array(
-				'key'   => 'trash',
-				'title' => __( 'Trash', 'tutor' ),
-				'value' => $trash,
-				'url'   => $url . '&data=trash',
-			),
+		$where = array();
+
+		if ( '' !== $date ) {
```
Comment (ahamed) [inline ecommerce/OrderController.php:197]: We can use `!empty($date)` that could check all kind of falsy values.

File: ecommerce/OrderController.php [hunk 5]
```diff
@@ -103,77 +181,49 @@ public function prepare_bulk_actions(): array {
...
+			$where['created_at_gmt'] = tutor_get_formated_date( 'Y-m-d', $date );
+		}
+
+		if ( '' !== $payment_status ) {
```
Comment (ahamed) [inline ecommerce/OrderController.php:201]: Same here.

File: helpers/QueryHelper.php [hunk 6]
```diff
@@ -471,58 +494,172 @@ public static function table_exists( $table ) {
...
-		$where_clause = !empty($where) ? 'WHERE ' . $obj->build_where_clause($where) : '';
-		
+
+		$where_clause = !empty($where) ? 'WHERE ' . self::build_where_clause($where) : '';
+
+		if (!empty($search)) {
+			$search_clause = self::build_like_clause( $search );
+			// foreach ($search as $column => $value) {
+			// 	$search_clauses[] = $wpdb->prepare("{$column} LIKE %s", '%' . $wpdb->esc_like($value) . '%');
+			// }
```
Comment (ahamed) [inline helpers/QueryHelper.php:526]: Remove this if this is not required. If you keep this for reference then add a `//@TODO: comment` for that.

File: includes/tutor-general-functions.php [hunk 7]
```diff
@@ -790,20 +790,23 @@ function tutor_js_date_format_against_wp() {
...
+	 *
+	 * @param string $require_format string If empty Y-m-d is used.
+	 * @param string $user_date string Date.
+	 *
+	 * @return string ( date )
+	 */
+	function tutor_get_formated_date( string $require_format = '', string $user_date ) {
+		if ( '' === $require_format ) {
+			$require_format = 'Y-m-d';
+		}
```
Comment (ahamed) [inline includes/tutor-general-functions.php:808]: We could use ternary shorthand.
```suggestion
		$require_format = $require_format ?: 'Y-m-d';
```

File: models/OrderModel.php [hunk 8]
```diff
@@ -5,47 +5,523 @@
...
+	 * @since 3.0.0
+	 *
+	 * @return array
+	 */
+	public static function get_payment_status() {
+		return array(
+			self::PAYMENT_STATUS_PENDING => __( 'Pending', 'tutor' ),
+			self::PAYMENT_STATUS_PAID    => __( 'Paid', 'tutor' ),
+			self::PAYMENT_STATUS_FAILED  => __( 'Failed', 'tutor' ),
+			self::PAYMENT_STATUS_UNPAID  => __( 'Unpaid', 'tutor' ),
```
Comment (ahamed) [inline models/OrderModel.php:119]: It seems logically `Pending` and `Unpaid` are same.

File: models/OrderModel.php [hunk 9]
```diff
@@ -5,47 +5,523 @@
...
+			'id'
+		);
+
+		if ( empty( $order_activities ) ) {
+			return array();
+		}
+
+		$response = array();
+
+		foreach ( $order_activities as &$activity ) {
```
Comment (ahamed) [inline models/OrderModel.php:344]: Don't need to use `&` here. 
```suggestion
		foreach ( $order_activities as $activity ) {
```

File: models/OrderModel.php [hunk 10]
```diff
@@ -5,47 +5,523 @@
...
+			'id'
+		);
+
+		if ( empty( $order_refunds ) ) {
+			return array();
+		}
+
+		$response = array();
+
+		foreach ( $order_refunds as &$refund ) {
```
Comment (ahamed) [inline models/OrderModel.php:404]: Same here.

File: models/OrderModel.php [hunk 11]
```diff
@@ -5,47 +5,523 @@
...
+				'type'  => 'LEFT',
+				'table' => "{$wpdb->users} u",
+				'on'    => 'o.user_id = u.ID',
+			),
+		);
+
+		$select_columns = array( 'o.*', 'u.user_login' );
+
+		$search_clause = array();
+		if ( '' !== $search_term ) {
```
Comment (ahamed) [inline models/OrderModel.php:477]:
```suggestion
		if ( !empty( $search_term ) ) {
```

**Review by ahamed — CHANGES_REQUESTED** [review]

**Pull request merged**